add Matern32 prim kernel

drezap opened the pull request:
## Summary

Adds the matern32 kernel primitive

## Tests

Tests column vectors and error handling.

## Side Effects

No side effects.

## Checklist

- [x] Math issue #864 

- [x] Copyright holder: (Copyright 2018 Andre Zapico)

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,278 @@
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern32_cov", "length scale", length_scale[n]);
+
+  check_size_match("gp_matern32_cov", "x dimension", x[0].size(),
```
This will fail in the same way as the other PR that was closed. If `x` is size 0, this will seg fault.

drezap commented:
coupled_ode_system unit tests fail, something to be worried about?

drezap commented:
@syclik

drezap commented:
Yo @syclik thanks for the reviews. Hope I’m not being too difficult. I appreciate you. As we all do.

Yeah, you’re right. I should be testing this better. so far, I’ve got: different datatypes for autodiff to work/templating, all of Stan’s different interfaces, obviously mathematical accuracy, trivial edge cases, invalid user input via the language... starting to accumulate these things thank you. I thought using the same test cases as before would be valid, but it’s not as good programmers make their code robust to bugs they don’t always test.

syclik commented:
@drezap, you're welcome! You're not... some of this stuff is hard to explain. Sorry, been sick for the last few days. I'm back on reviewing all this.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,279 @@
+ *
+ * @param x1 std::vector of Eigen column vectors of scalars
+ * @param x2 std::vector of Eigen column vectors of scalars
+ *    This function assumes each column of x1 and x2 are the same size.
```
Please remove this doc. We don't assume that `x1` and `x2`'s columns have the same length.

syclik replied:
Did you see this comment? It still holds. Doc is incorrect here.

drezap commented:
@syclik  yo - just wanna make sure I'm being clear. I'm going to explain this in a really naive way, but we're interested in different covariance functions because with different covariance functions, we can impose different properties on the process. They can be any number of things, smoothness, periodic properties, etc. Which is why the ultimate goal is to let users implement their own. Anyway - I'm implementing pretty much exactly as it is on the wikipedia page/GPML (with some exceptions to get it to work with C++ and stan's language). With that said, we have the first two for "in sample data"/training data, what ever you wanna call it. Second two with two input data for the out of sample data. in the multivariate gaussian mean function, this is f = K_{x*, x} (K_{x,x}^{-1}) y, for a posterior predictive with no noise (2.19 in GPML).

This really shouldn't need any maintenance moving forward, as the code is pretty f* simple, but if it does need maintenance, it's copy written to me, and I take this seriously, and if you ping me I'll come fix it

syclik commented:
Just a quick thing to make it easier to review: can you merge in develop? I feel like I'm looking at changes on GitHub that aren't relevant.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,285 @@
+
+  T_s sigma_sq = square(sigma);
+  T_l root_3 = sqrt(3.0);
+  T_l neg_root_3 = -1.0 * sqrt(3.0);
```
For efficiency, can you declare `root_3` and `neg_root_3` as `double`? As written, this will promote these to type `T_l`.

drezap commented:
@syclik yup sorry. pulled from develop branch and now you should only see the 3 files I've edited. Thanks

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,284 @@
+  check_size_match("gp_matern32_cov", "x2 dimension", x2[0].size(),
+                   "number of length scales", l_size);
+
+  double sigma_sq = square(sigma);
```
This isn't correct any more. The tests shouldn't allow this to compile. The type of `sigma_sq` needs to be `T_s`. There are a handful of these problems in your latest merge. 

`sqrt(3.0)` should be `double` because we know that type exactly.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,284 @@
+ *
+ * @param x1 std::vector of scalars that can be used in squared_distance
+ * @param x2 std::vector of scalars that can be used in squared_distance
+ *    This function assumes all elements of x1 and x2 are the same length.
```
Can we clarify this? I think we should assume all elements of x1 has the same size. Same with X2, but not them both. I'm thinking users from the Stan language can make that mistake. 

Can we add the checks from the other PR where we check that the size of the elements of x1 matches X2?

drezap replied:
from the other PR, we check where the size matches x1 and x2 implicitly when we check to see that the number of of dimensions are same as 'P' (or D) for each x. I agree, and I'll go ahead and add checks for all, so the user doesn't make this mistake

drezap replied:
Ok, so here I'm encountering issues, and I have reason to believe this is why the assumption was here in the first place. my compiler is not happy trying to get the `size()` of an element when it's a scalar (because T_x can either be a scalar type or vector type for the 1st and 3rd functions... so I have to do some tedious metaprogramming or something)

syclik replied:
We have a template meta program for that! I don't know what it is off the top of my head, but it's there! Don't rewrite it unless we really don't have it.

syclik replied:
We have a template meta program for that! I don't know what it is off the top of my head, but it's there! Don't rewrite it unless we really don't have it.

drezap commented:
@syclik So here's a summary of the changes in this commit: fixed all the templating and using `return_type` where appropriate. I now check _every element_ of x, or x1, x2 for the cross covariance calculation to see if it's the same size. I'm also using `stan::size_of` because this is much safer. I've also added relevant test cases.

For the ARD case, I'm only checking to see if the first element is the same size as number of length scales. If you also want me to parse through all elements of x to check the length I can do so, just let me know.

drezap commented:
I just realized that this size check is done implicitly in `squared_distance`, though. So it's not even that necessary to make sure they're the same size, but the errors could serve a pedagogical purpose.

syclik commented:
Thanks for adding these changes. It looks like some stuff got lost in translation. I'll have to review from a computer, but I'll try to provide a lot more specific guidance.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+    return cov;
+
+  const char *function = "gp_matern32_cov";
+  for (size_t i = 0; i < x_size; ++i)
```
This logic doesn't seem right. It'll do the right thing, but it's really wasteful. `x_size * (x_size - 1) / 2` times, but you should only need to make `x_size` comparisons.

So, first off, it's perfectly acceptable to document this to say `@param x std::vector of scalars that can be used in square distance. It's assumed each element of x is of the same size`. Then we could just check the first element. If you don't want to do that, the logic here should be something like:
```
size_t x_element_size = size_of(x[0]);
for (size_t i = 1; i < x_size; ++i)
  check_size_match(function, "x row", size_of(x[i]), "x's other row", x_element_size);
```

Since we've already checked that `x_size > 0` (line 51), this is ok to do this way.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+  T_s sigma_sq = square(sigma);
+  typename return_type<double, T_l>::type root_3_inv_l
+      = sqrt(3.0) / length_scale;
+  typename return_type<double, T_l>::type neg_root_3_inv_l
```
This type should be `T_l` also.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+  for (size_t i = 0; i < x_size; ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      typename return_type<T_x, T_s, T_l>::type distance
```
This type is incorrect. It should be `T_x`.

`return_type<T_x, T_s, T_l>::type` will be `stan::math::var` if any of the three are `stan::math::var`. You'll be getting overpromotion if `T_x` is `double`. Meaning if `T_x` is actually `double`, then the `sqrt(squared_distance(x[i], x[j)))` will also be `double`, but if `T_s` happens to be `stan::math::var`, we've now constructed an autodiff variable and placed it on the stack. So please change this to `T_x`.

syclik replied:
[question]
do we not have a `distance` function?

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+
+  for (size_t i = 0; i < x_size; ++i) {
+    for (size_t j = i; j < x_size; ++j) {
+      typename return_type<T_x, T_s, T_l>::type distance
```
This type is incorrect. It should be `return_type<T_x, T_l>::type`.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+    return cov;
+  const char *function = "gp_matern32_cov";
+
+  for (size_t i = 0; i < x1_size; ++i)
```
Similar comment to that above. This does the right thing, but is inefficient.

You can either document that you assume all of `x1`'s elements are the same size and the same for `x2`. But this has to be consistent with the other function. Or... you should do this with two loops, the first over `x1_size` and the second over `x2_size`.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+  check_positive_finite(function, "length scale", length_scale);
+
+  T_s sigma_sq = square(sigma);
+  typename return_type<double, T_l>::type root_3_inv_l_sq
```
Change type to `T_l`.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+  T_s sigma_sq = square(sigma);
+  typename return_type<double, T_l>::type root_3_inv_l_sq
+      = sqrt(3.0) / length_scale;
+  typename return_type<double, T_l>::type neg_root_3_inv_l_sq
```
Change type to `T_l`.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+
+  for (size_t i = 0; i < x1_size; ++i) {
+    for (size_t j = 0; j < x2_size; ++j) {
+      typename return_type<T_x1, T_x2>::type distance
```
[comment]

This is the correct type. The result is `stan::math::var` if either `T_x1` or `T_x2` is `stan::math::var`, but `double` otherwise.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+  for (size_t n = 0; n < l_size; ++n)
+    check_not_nan(function, "length scale", length_scale[n]);
+
+  check_size_match(function, "x1 dimension", size_of(x1[0]),
```
This function is inconsistent with the rest. The others make sure that all elements of `x1` are the same length and do not assume they are. Here, this only checks `x1[0]` and not `x1[1], ..., x1[x1_size-1]`.

syclik commented on stan/math/prim/mat/fun/gp_matern32_cov.hpp:
```diff
@@ -0,0 +1,304 @@
+
+  for (size_t i = 0; i < x1_size; ++i) {
+    for (size_t j = 0; j < x2_size; ++j) {
+      typename return_type<T_x1, T_x2, T_s, T_l>::type distance
```
Incorrect type. This should be `return_type<T_x1, T_x2, T_l>::type`.

syclik commented:
> I just realized that this size check is done implicitly in `squared_distance`, though. So it's not even that necessary to make sure they're the same size, but the errors could serve a pedagogical purpose.

It really does help to provide good error messages. Can you imagine if we just passed error messages through like that? Calling `y ~ normal(mu, sigma);` could end up with an error message that says the problem is in `divide` and it'll really be difficult to debug.

One thing you could do... wrap it in a try-catch block and throw an appropriate error if it fails. We haven't used that pattern in our code base, but I don't see why it couldn't be done that way.

drezap commented:
If people are confused I can see that being a good idea. I'm opting out of using the word "dimension" because it's ambiguous, and I'm sticking saying "number of elements in a row" (which means number covariates, columns, etc)

drezap commented:
Yo @syclik - can we keep this moving? I’ll still keep the flatten going, but I like to work on a 1-3 different things at once, and I’d like to get this in. There’s 5-6 more of these, but I think this is getting close to being ready for user- use

syclik commented:
@drezap: yup. Just FYI, the reviewer doesn't get notified when the code changes. So ping me whenever you get to that point.
